### PR TITLE
Add documentation for external code generation

### PR DESCRIPTION
**Pull request by jpeeler**
This document is a starting point for those wishing to utilize the existing
code generation abilities of Kubernetes in an external project.

### CODE REVIEW

**Comment by jpeeler** [comment]
Feedback/corrections very welcome.

@pmorie @DirectXMan12 @sjenning

**Comment by pmorie** [comment]
cc @deads2k @lavalamp

File: contributors/devel/code-generation-outside-kubernetes.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+Each code generation binary supports different arguments. The arguments of
+importance are:
+--extra-peer-dirs - defines additional locations to provide current operations
+--bounding-dirs - list of import paths for binding the types
+--input-dirs - list of import paths to get input types from
+
+In service-catalog, the makefile has been significantly simplified and one
+can see the parameters sent to each code generator here:
+
+https://github.com/kubernetes-incubator/service-catalog/blob/59cd78081f6cb23aa102df493f5b2471e8e0cdd2/Makefile#L102
```
Comment (chrislovecnm) [inline contributors/devel/code-generation-outside-kubernetes.md]: Can we pull in examples?  For instance the `service-catalog` is amazing and is split into three different projects.  Now we have a doc link that needs updates.  Just a thought.

File: contributors/devel/code-generation-outside-kubernetes.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+generation behavior. These options are described in either main.go for a
+given generator or within gengo itself.
+
+## Potential pitfall
+When importing Kubernetes api it's important to remember to import the
+correct version of the api. Always align the api with the same version
+package that is being built:
+
+https://github.com/kubernetes-incubator/service-catalog/blob/59cd78081f6cb23aa102df493f5b2471e8e0cdd2/pkg/apis/servicecatalog/types.go#L20
+https://github.com/kubernetes-incubator/service-catalog/blob/59cd78081f6cb23aa102df493f5b2471e8e0cdd2/pkg/apis/servicecatalog/v1alpha1/types.go#L20
```
Comment (chrislovecnm) [inline contributors/devel/code-generation-outside-kubernetes.md]: Can we have a TLDR; ??

So you want to write api

1. write this struct, with this layout
2. generate the apimachinery code
3. build you binaries
4. package it this way.

@pmorie what do you think of a layout like that?

File: contributors/devel/code-generation-outside-kubernetes.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
+# Code generation (from the experience of developing service-catalog)
```
Comment (chrislovecnm) [inline contributors/devel/code-generation-outside-kubernetes.md]: Super picky English observation.  You are writing a lot like me ... in passive voice ;)  I really try to write in active voice :)  Some cool online tools which can assist if you are interested.

File: contributors/devel/code-generation-outside-kubernetes.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+## Code generator binaries
+There are several libraries in the Kubernetes tree that are responsible for
```
Comment (chrislovecnm) [inline contributors/devel/code-generation-outside-kubernetes.md]: > Kubernetes tree

Playing dumb n00b ... what is the K8s tree?

File: contributors/devel/code-generation-outside-kubernetes.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+various code generation operations. Those go based libraries are currently
+located at cmd/libs/go2idl/. In each directory, there is a main.go file
+that currently has hard coded an internal path for boilerplate.go.txt.
+However, there now exists a repo-infra repository that can be used instead.
```
Comment (chrislovecnm) [inline contributors/devel/code-generation-outside-kubernetes.md]: > repo-infra repository

huh? Where is it

File: contributors/devel/code-generation-outside-kubernetes.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+These files serve mostly as wrappers for executing and utilizing gengo.
+
+## Code generation arguments
+In Kubernetes, most of the code generation execution handling is in the
```
Comment (chrislovecnm) [inline contributors/devel/code-generation-outside-kubernetes.md]: > In Kubernetes

Should we a bit more specific?  K8s is HUGE.  Where in K8s?

File: contributors/devel/code-generation-outside-kubernetes.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+Makefile.generated_files and is in the .generated_files target. The rule
+complexity is due to Kubernetes generating a DAG in order to determine
+which files need rebuilt and code regenerated for.
```
Comment (chrislovecnm) [inline contributors/devel/code-generation-outside-kubernetes.md]: I am fairly new to apimachinery so

> The rule complexity is due to Kubernetes generating a DAG in order to determine which files need rebuilt and code regenerated for.

I need a lot more context, and ELI5.

> need rebuilt

Incorrect verb form

File: contributors/devel/code-generation-outside-kubernetes.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+--bounding-dirs - list of import paths for binding the types
+--input-dirs - list of import paths to get input types from
+
+In service-catalog, the makefile has been significantly simplified and one
+can see the parameters sent to each code generator here:
+
+https://github.com/kubernetes-incubator/service-catalog/blob/59cd78081f6cb23aa102df493f5b2471e8e0cdd2/Makefile#L102
+
+Special note is the conversion-gen arguments, which include Kubernetes
+directories. The reason for this is because the Kubernetes types have
```
Comment (chrislovecnm) [inline contributors/devel/code-generation-outside-kubernetes.md]: > Special note is the conversion-gen arguments, which include Kubernetes directories.

The noun and verb seem a bit odd ... "A special note about xyz??"

File: contributors/devel/code-generation-outside-kubernetes.md (hunk 2 above)
Comment (chrislovecnm) [inline contributors/devel/code-generation-outside-kubernetes.md]: Still a bit fuzzy on ... I have my codegen ... does that create a binary?  How does the HTTP server actually start?  I have cobra cli now I need that to start something.

File: contributors/devel/code-generation-outside-kubernetes.md (hunk 3 above)
Comment (jpeeler) [inline contributors/devel/code-generation-outside-kubernetes.md]: I understand what you're saying, but don't know how to rewrite. Suggestion?

File: contributors/devel/code-generation-outside-kubernetes.md (hunk 4 above)
Comment (jpeeler) [inline contributors/devel/code-generation-outside-kubernetes.md]: I'll reword as "repo", that's a better word anyway.

File: contributors/devel/code-generation-outside-kubernetes.md (hunk 5 above)
Comment (jpeeler) [inline contributors/devel/code-generation-outside-kubernetes.md]: https://github.com/kubernetes/repo-infra, I'll update

File: contributors/devel/code-generation-outside-kubernetes.md (hunk 6 above)
Comment (jpeeler) [inline contributors/devel/code-generation-outside-kubernetes.md]: I've added the directory of the code generators, which are in cmd/libs/go2idl.

File: contributors/devel/code-generation-outside-kubernetes.md (hunk 7 above)
Comment (jpeeler) [inline contributors/devel/code-generation-outside-kubernetes.md]: Fixed verb. I wish I understood all the makefile dependency generation better...

File: contributors/devel/code-generation-outside-kubernetes.md (hunk 8 above)
Comment (jpeeler) [inline contributors/devel/code-generation-outside-kubernetes.md]: fixed

File: contributors/devel/code-generation-outside-kubernetes.md (hunk 1 above)
Comment (jpeeler) [inline contributors/devel/code-generation-outside-kubernetes.md]: I don't understand what you mean by three different projects, but the makefile section that I was linking to above is practically the same today. I tend to like the linking here, but can pull the makefile content in if that would help understanding.

**Comment by bgrant0607** [comment]
cc @kubernetes/sig-api-machinery-pr-reviews
